### ISSUE TITLE
Built-in TS support incorrectly reporting problems on fetch headers.set()

### ISSUE THREAD

**Issue by mjbvz** [Microsoft/TypeScript#23383]
_From @nickdepinet on April 11, 2018 22:54_

Version 1.22.1
Commit 950b8b0d37a9b7061b6f0d291837ccc4015f5ecd
Date 2018-04-06T02:26:57.615Z
Shell 1.7.12
Renderer 58.0.3029.110
Node 7.9.0
Architecture x64
OS: windows 10
Steps to Reproduce:

I have a typescript project using the es6 fetch apis. In certain files I use the following:

![image](https://user-images.githubusercontent.com/2041026/38647142-b0c2cff0-3d9f-11e8-8c80-c10eee27ca8c.png)

The code builds and runs correctly, but vscode reports problems in the files

![image](https://user-images.githubusercontent.com/2041026/38647251-231f83ea-3da0-11e8-94d0-4fdfb020b248.png)

Digging down in it appears that code is only checking the first possible type on HeadersInit:

![image](https://user-images.githubusercontent.com/2041026/38647308-6f5e38e6-3da0-11e8-82b9-f49fc3078039.png)

Does this issue occur when all extensions are disabled?: Yes

_Copied from original issue: Microsoft/vscode#47701_

**Comment by mjbvz** [comment]
Error seems to be that `options.headers` is of type `HeadersInit` while `.set` only exists on the `Headers` type. Not sure if the types are correct or not

**Comment by mhegazy** [comment]
Where is `options.headers` defined?

**Comment by nickdepinet** [comment]
here's the whole start of the function in question: 

![image](https://user-images.githubusercontent.com/2041026/38713455-119a9998-3e86-11e8-96cd-a52d094c80d4.png)

`options` is occassionally passed in with `options.headers` being an instance of `Headers`. otherwise it is set in the function as above - it's never set as a `string[][]`.

It doesn't appear correct for vscode to always assume `HeadersInit` is a `string[][]` instead of the other possible options, which have `set()` as a function on them:

![image](https://user-images.githubusercontent.com/2041026/38713685-45922170-3e87-11e8-9dcd-159214d8dd91.png)

**Comment by mhegazy** [comment]
The underlying issue here is that `options` is declared to have the type `RequestInit`. So the input for your function can be `{ headers: { foo: 2}}`, which does not have `set` function on it at runtime.

e.g.:

```ts
function sendRequest(options?: RequestInit) {
    options = options || {};
    const headers = options.headers || new Headers();
    headers.set()
}

sendRequest({ headers: { foo: "" } }); // OK, but fails at runtime.

```

Also for future references, copy/past code is much easier than screenshots..

**Comment by nickdepinet** [comment]
The error messaging here could definitely be more clear, you can be operating entirely in a functional scenario and still get reported problems.
